py/server/deephaven/barrage.py

```diff
@@ -0,0 +1,180 @@
+             as True. Defaults to None
+
+    Returns:
+        a Deephaven Barrage session
+
+    Raises:
+        DHError
+    """
+    try:
+        if tls_root_certs and not use_tls:
+            raise DHError(message="tls_root_certs is provided but use_tls is False")
+
+        target_uri = f"{host}:{port}"
+        if use_tls:
+            target_uri = f"dh://{target_uri}"
+        else:
+            target_uri = f"dh+plain://{target_uri}"
+
+        j_client_config = _build_client_config(target_uri, tls_root_certs)
+        auth = f"{auth_type} {auth_token}"
+
+        try:
+            return _get_barrage_session_uri(j_client_config, auth)
+        except:
+            # fall back to the direct way when we don't have a fully initialized server, used for testing
+            return _get_barrage_session_direct(j_client_config, auth)
+    except Exception as e:
+        raise DHError(e, "failed to get a barrage session to the target remote Deephaven server.") from e
+
+
+def _get_barrage_session_uri(client_config, auth) -> BarrageSession:
+    j_barrage_session_factory_client = uri.resolve(
+        "dh:///app/io.deephaven.server.barrage.BarrageSessionFactoryClient/field/instance")
+    j_barrage_session_factory = j_barrage_session_factory_client.factory(client_config, auth)
+    return BarrageSession(j_barrage_session_factory.newBarrageSession())
+
+
+def _get_barrage_session_direct(client_config, auth) -> BarrageSession:
+    """Note, this is used for testing only. This way of constructing a Barrage session is less efficient because it does
+    not share any of the state or configuration that the server provides; namely, when you are doing it with the server
+    context it provides a singleton executor, allocator, outbound SSL configuration, and the ability for the server to
+    hook in additional channel building options.
+    """
+    j_channel = _JChannelHelper.channel(client_config)
+    j_dh_channel = _JDeephavenChannelImpl(j_channel)
+
+    j_session_config = (_JSessionImplConfig.builder()
+                        .executor(_JExecutors.newScheduledThreadPool(4))
+                        .authenticationTypeAndValue(auth)
+                        .channel(j_dh_channel)
+                        .build())
+    j_session = _JSessionImpl.create(j_session_config)
+    return BarrageSession(_JBarrageSession.create(j_session, _JRootAllocator(), j_channel))
+
+
+def _build_client_config(target_uri, tls_root_certs) -> jpy.JType:
+    j_client_config_builder = _JClientConfig.builder()
+    j_client_config_builder.target(_JDeephavenTarget.of(_JURI(target_uri)))
+    if tls_root_certs:
+        j_ssl_config = _JSSLConfig.builder().trust(
+            _JTrustCustom.ofX509(tls_root_certs, 0, len(tls_root_certs))).build()
+        j_client_config_builder.ssl(j_ssl_config)
+    j_client_config = j_client_config_builder.build()
+    return j_client_config
+
+
+class BarrageSession(JObjectWrapper):
```

Comment:
code would be easier to read if this was at the top of the file, before the functions that return it.